dev/src/document-group.ts

```diff
@@ -0,0 +1,348 @@
+   * let docB = firestore.doc('coll/b');
+   * let documentGroup = firestore.documentGroup(docA, docB);
+   *
+   * let count = 0;
+   *
+   * query.stream().on('data', documentSnapshot => {
+   *   console.log(`Retrieved document with name '${documentSnapshot.id}'`);
+   *   ++count;
+   * }).on('end', () => {
+   *   console.log(`Total count is ${count}`);
+   * });
+   */
+  stream(): NodeJS.ReadableStream {
+    const uniqueDocuments =
+        removeDuplicates(this.documents.map(docRef => docRef.formattedName));
+
+    const request: api.IBatchGetDocumentsRequest = {
+      database: this.firestore.formattedName,
+      transaction: this.options.transactionId,
+      documents: uniqueDocuments
+    };
+
+    if (this.options.fieldMask) {
+      const fieldPaths =
+          this.options.fieldMask.map(fieldPath => fieldPath.formattedName);
+      request.mask = {fieldPaths};
+    }
+
+    const tag = requestTag();
+    const responseStream: NodeJS.ReadStream = through2.obj();
+
+    this._firestore.readStream('batchGetDocuments', request, tag, true)
+        .then(backendStream => {
+          backendStream.on('error', err => {
+            logger(
+                'DocumentGroup.stream', tag,
+                'BatchGetDocumentsRequest failed with stream error:', err);
+            responseStream.destroy(err);
+          });
+          backendStream.on(
+              'data', (response: api.BatchGetDocumentsResponse) => {
+                let snapshot: DocumentSnapshot;
+
+                try {
+                  if (response.found) {
+                    logger(
+                        'DocumentGroup.stream', tag, 'Received document: %s',
+                        response.found.name!);
+                    snapshot = this._firestore.snapshot_(
+                        response.found, response.readTime);
+                  } else {
+                    logger(
+                        'DocumentGroup.stream', tag, 'Document missing: %s',
+                        response.missing);
+                    snapshot = this._firestore.snapshot_(
+                        response.missing, response.readTime);
+                  }
+                  responseStream.push(snapshot);
+                } catch (err) {
+                  logger(
+                      'DocumentGroup.stream', tag,
+                      'Failed to deserialize document: %s', err);
+                  backendStream.end();
+                  responseStream.destroy(err);
+                }
+              });
+          backendStream.on('end', () => {
+            responseStream.end();
+          });
+          backendStream.resume();
+        })
+        .catch(err => {
+          responseStream.destroy(err);
+        });
+
+
+    return responseStream;
+  }
+
+  /**
+   * Attaches a listener to retrieve real-time change events for the document
+   * in this document group.
+   *
+   * The results are ordered by document name and missing documents are
```

Comment:
Decision: .doc() should return null if the document isn't in the QuerySnapshot (even though technically it might exist but they didn't include it in their .documents() list.